[10.0] sale promotion rule:  Supports discount amount

lmignon opened the pull request:
- [X] includes #729  

Before this change, only percentage were applicable as discount by promotion rule.
This change allows you to define rules that apply a discount in the form of a specific amount.

lmignon committed:
[IMP] sale_promotion_rule: Supports discount amount
Before this change, only percentage were applicable as discount by
promotion rule.
This change allows you to define rules that apply a discount in the form of a specific amount. This amount is translated as a percentage disctributed on the different lines during the rule resolution in order to benefit from the standard mechanism for calculating discount by Odoo.

lmignon commented:
CC @sebastienbeau @Cedric-Pigeon

lmignon commented:
CC @bguillot

nikul-serpentcs commented on sale_promotion_rule/data/sale_promotion_rule_data.xml:
```diff
@@ -0,0 +1,8 @@
+    <record id="product.decimal_discount" model="decimal.precision">
+        <field name="digits">3</field>
+    </record>
+</odoo>
```
@lmignon add a new blank line

nikul-serpentcs commented on sale_promotion_rule/models/sale_order.py:
```diff
@@ -0,0 +1,64 @@
+    applied_promotion_rule_ids = fields.Many2many(
+        'sale.promotion.rule',
+        string='Promotion rules',
+        compute='_compute_applied_promotion_rule_ids'
```
@lmignon IMO change `_compute_has_promotion_rules`

lmignon replied:
@nikul-serpentcs IMO we must keep one compute method by field to ease readability, maintainability and possible extension. An all in one method is never a good idea.

nikul-serpentcs commented on sale_promotion_rule/models/sale_order.py:
```diff
@@ -0,0 +1,64 @@
+                rec.promotion_rule_ids)
+
+    @api.depends('promotion_rule_ids', 'coupon_promotion_rule_id')
+    def _compute_applied_promotion_rule_ids(self):
```
@lmignon Code merge in this  `_compute_has_promotion_rules` Method,

Because here same depends `@api.depends('promotion_rule_ids', 'coupon_promotion_rule_id')`
no need to new method

lmignon replied:
No see my previous comment.

nikul-serpentcs commented on sale_promotion_rule/models/sale_order_line.py:
```diff
@@ -0,0 +1,50 @@
+                rec.promotion_rule_ids)
+
+    @api.depends('promotion_rule_ids', 'coupon_promotion_rule_id')
+    def _compute_applied_promotion_rule_ids(self):
```
@lmignon same as for this method also

lmignon replied:
No, see my previous comment.

nikul-serpentcs reviewed (commented):
Improve Code

lmignon committed:
[IMP] sale_promotion_rule: Use discount product for discount defined with fixed amount
While the first approach based on the reuse of the standard Odoo field "discount_amount" seemed the least intrusive, it is unfortunately impossible to avoid rounding problems due to the transformation of a fixed amount into a percentage of dicount to be applied on the different products. Rounding problems are further aggravated when a so contains lines with products with a high price difference.
Since it's not possible to rely on the standard Odoo field, the implementation of the fixed amount discount is now based the usage of a dedicated product.

lmignon commented:
ping @Cedric-Pigeon

lmignon committed:
[IMP] sale_promotion_rule: Improve tests coverage

lmignon committed:
[FIX] sale_promotion_rule: Fix search for rules already used for one partner

rousseldenis commented on sale_promotion_rule/models/sale_promotion_rule.py:
```diff
@@ -95,6 +105,37 @@ class SalePromotionRule(models.Model):
         ('code_unique', 'UNIQUE (code)', _('Discount code must be unique !'))]
 
+    @api.constrains("discount_product_id", "promo_type", "discount_type")
+    def _validate_promotion_product_id(self):
```
Please follow https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#models for method name.

Same here under

rousseldenis commented on sale_promotion_rule/models/sale_promotion_rule.py:
```diff
@@ -95,6 +105,37 @@ class SalePromotionRule(models.Model):
+                    "applaying a specific amount"
+                ))
+
+    @api.constrains("promo_type", "discount_type")
```
Missing discount_amount_currency_id

rousseldenis reviewed (changes requested):
Minor comments

rousseldenis commented:
@nikul-serpentcs Could you update your review ?

@lmignon Could you attend comments ?

@Cedric-Pigeon @acsonefho

Commit:
[FIX] prevent to aply an automatic rule on a sale promotion line

sebastienbeau commented on sale_promotion_rule/models/sale_order.py:
```diff
@@ -13,13 +13,15 @@ class SaleOrder(models.Model):
         'sale.promotion.rule',
         string='Promotion rules',
         domain=[('rule_type', '!=', 'coupon')],
+        inedx=True,
```
Typo error

sebastienbeau commented on sale_promotion_rule/data/sale_promotion_rule_data.xml:
```diff
@@ -0,0 +1,8 @@
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo>
+    <record id="product.decimal_discount" model="decimal.precision">
+        <field name="digits">3</field>
```
@lmignon why do you change the decimal precision of the discount?
This will change the default value and if the module is installed break existing test in other module

sebastienbeau reviewed (commented):
Great addition, minor comment

lmignon commented:
@rousseldenis @sebastienbeau All your comments are addressed. Can we move forward?

lmignon commented:
travis is :red_circle: due to https://github.com/OCA/sale-workflow/issues/817

acsonefho commented:
Fro the info, I just did a rebase to avoid errors on MR https://github.com/shopinvader/odoo-shopinvader/pull/313

rafaelbn commented:
Please @rousseldenis could your review this PR of @lmignon ?

rousseldenis commented:
All :green_apple: ! @rafaelbn

rousseldenis committed:
!fixup

rafaelbn reviewed (approved):
:+1:

rafaelbn commented:
/ocabot merge patch
